### ISSUE TITLE
Lazy execution of scope_items for connection types raises an exception

### ISSUE THREAD

**Issue by gumchum** [rmosolgo/graphql-ruby#2747]
I'm running into an issue involving `scope_items`, connections, and lazy execution using promises.

I've defined the `scope_items` hook on my schema object type to return a promise:

```
Post = Struct.new(:id, :name) 

class PostType < GraphQL::Schema::Object 
  field :id, ID, null: false             
  field :name, String, null: false       
                                         
  def self.scope_items(items, context)   
    Promise.resolve(items)               
  end                                    
end
```

These objects are then returned through a connection:

```
class QueryType < GraphQL::Schema::Object                                 
  field :posts, PostType.connection_type, null: false                     
                                                                          
  def posts                                                               
    [Post.new(1, 'Post 1'), Post.new(2, 'Post 2'), Post.new(3, 'Post 3')] 
  end                                                                     
end                                                                       
```

```
class MySchema < GraphQL::Schema
  query QueryType               
  lazy_resolve Promise, :sync   
end                             
```

Execution of a query using this connection field results in an exception:

```
/Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/relay/base_connection.rb:36:in `connection_for_nodes': No connection implementation to wrap Promise (#<Promise:0x00007fe8219d9ea8>) (RuntimeError)
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field/connection_extension.rb:35:in `after_resolve'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:663:in `block (2 levels) in with_extensions'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:660:in `each'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:660:in `each_with_index'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:660:in `block in with_extensions'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:1206:in `after_lazy'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:659:in `with_extensions'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:623:in `public_send_field'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:481:in `block in resolve_field'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:1206:in `after_lazy'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/field.rb:465:in `resolve_field'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/member/instrumentation.rb:80:in `call'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/member/instrumentation.rb:80:in `call'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/field.rb:248:in `resolve'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:321:in `call'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/middleware_chain.rb:49:in `invoke_core'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema/middleware_chain.rb:38:in `invoke'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:129:in `block in resolve_field'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `block in trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:76:in `call_tracers'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:128:in `resolve_field'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:92:in `block in resolve_selection'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:85:in `each'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:85:in `resolve_selection'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:56:in `block in resolve_root_selection'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `block in trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:76:in `call_tracers'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:49:in `resolve_root_selection'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/execute.rb:31:in `begin_query'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:112:in `begin_query'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:84:in `block in run_as_multiplex'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:83:in `map'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:83:in `run_as_multiplex'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:62:in `block (2 levels) in run_queries'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:186:in `block in instrument_and_analyze'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:174:in `instrument_and_analyze'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:61:in `block in run_queries'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `block in trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:76:in `call_tracers'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/tracing.rb:62:in `trace'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:59:in `run_queries'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/execution/multiplex.rb:49:in `run_all'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:392:in `block in multiplex'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:1279:in `with_definition_error_check'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:391:in `multiplex'
        from /Users/brianchhun/.rbenv/versions/2.4.3/lib/ruby/gems/2.4.0/gems/graphql-1.9.14/lib/graphql/schema.rb:368:in `execute'
        from scopeitems.rb:32:in `<main>'
```

Things seem to work fine if I use `[PostType]` instead of using a connection.

**Comment by rmosolgo** [comment]
Thanks for reporting this issue! It sounds like it's missing an `after_lazy(...)` call, which would tell it to wait for that promise.

**Comment by gumchum** [comment]
Thanks @rmosolgo. Can you clarify what you mean by that? I'm guessing the solution is to handle the lazy objects here: https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/field.rb#L718

**Comment by rmosolgo** [comment]
Sorry I didn't follow up on your question before 😖 ! This will be fixed in 1.11.3 by #3070, but it requires opting into the [new runtime](https://graphql-ruby.org/queries/interpreter.html). 

That runtime will be the default in 1.12.0, so it's worth migrating if you can. But if you can't, we could probably backport the change to an old graphql-ruby version, too. (Or you could use the diff as a monkey-patch!)
